src/lib.rs

```diff
@@ -104,58 +121,51 @@ impl<D> EventSource for WaylandSource<D> {
         &mut self,
-        readiness: Readiness,
-        token: Token,
+        _: Readiness,
+        _: Token,
         mut callback: F,
     ) -> Result<PostAction, Self::Error>
     where
         F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
     {
         let queue = &mut self.queue;
-        let read_guard = &mut self.read_guard;
+        if let Some(err) = self.stored_error.take() {
+            return Err(err)?;
+        }
 
-        let action = self.fd.process_events(readiness, token, |_, _| {
-            // 1. read events from the socket if any are available
-            if let Some(guard) = read_guard.take() {
-                // might be None if some other thread read events before us, concurrently
-                if let Err(WaylandError::Io(err)) = guard.read() {
-                    if err.kind() != io::ErrorKind::WouldBlock {
-                        return Err(err);
-                    }
-                }
-            }
+        // In theory, we would need to call the process_events handler on fd. However,
+        // we know that Generic's process events's call is a no-op, so we can just handle the event ourselves
 
-            // 2. dispatch any pending events in the queue
-            // This is done to ensure we are not waiting for messages that are already in
-            // the buffer.
-            Self::loop_callback_pending(queue, &mut callback)?;
-            *read_guard = Some(Self::prepare_read(queue)?);
-
-            // 3. Once dispatching is finished, flush the responses to the compositor
-            if let Err(WaylandError::Io(e)) = queue.flush() {
-                if e.kind() != io::ErrorKind::WouldBlock {
-                    // in case of error, forward it and fast-exit
-                    return Err(e);
-                }
-                // WouldBlock error means the compositor could not process all
-                // our messages quickly. Either it is slowed
-                // down or we are a spammer. Should not really
-                // happen, if it does we do nothing and will flush again later
-            }
+        // 2. dispatch any pending events in the queue
+        // This is done to ensure we are not waiting for messages that are already in
+        // the buffer.
+        Self::loop_callback_pending(queue, &mut callback)?;
 
-            Ok(PostAction::Continue)
-        })?;
+        // 3. Once dispatching is finished, flush the responses to the compositor
+        if let Err(WaylandError::Io(e)) = queue.flush() {
+            if e.kind() != io::ErrorKind::WouldBlock {
+                // in case of error, forward it and fast-exit
+                return Err(e.into());
+            }
+            // WouldBlock error means the compositor could not process all
+            // our messages quickly. Either it is slowed
+            // down or we are a spammer. Should not really
+            // happen, if it does we do nothing and will flush again later
+        }
```

Comment:
Should probably extract into its own method given that it's used in multiple places.